candle-transformers/src/models/stable_diffusion/uni_pc.rs

```diff
@@ -0,0 +1,1048 @@
+            let len = self.len() as usize;
+            debug_assert!(len < 5);
+            sort_floats(&mut heights[..len]);
+            let desired_index = (len as f64) * self.p() - 1.;
+            let mut index = desired_index.ceil();
+            if desired_index == index && index >= 0. {
+                let index = index.round() as usize;
+                debug_assert!(index < 5);
+                if index < len - 1 {
+                    return 0.5 * self.q[index] + 0.5 * self.q[index + 1];
+                }
+            }
+            index = index.max(0.);
+            let mut index = index.round() as usize;
+            debug_assert!(index < 5);
+            index = index.min(len - 1);
+            self.q[index]
+        }
+
+        fn len(&self) -> u64 {
+            self.n[4] as u64
+        }
+
+        fn is_empty(&self) -> bool {
+            self.len() == 0
+        }
+
+        pub fn add(&mut self, x: f64) {
+            self.max = self.max.map(|y| y.max(x)).or(Some(x));
+
+            if self.n[4] < 5 {
+                self.q[self.n[4] as usize] = x;
+                self.n[4] += 1;
+                if self.n[4] == 5 {
+                    sort_floats(&mut self.q);
+                }
+                return;
+            }
+
+            let mut k: usize;
+            if x < self.q[0] {
+                self.q[0] = x;
+                k = 0;
+            } else {
+                k = 4;
+                for i in 1..5 {
+                    if x < self.q[i] {
+                        k = i;
+                        break;
+                    }
+                }
+                if self.q[4] < x {
+                    self.q[4] = x;
+                }
+            };
+
+            for i in k..5 {
+                self.n[i] += 1;
+            }
+            for i in 0..5 {
+                self.m[i] += self.dm[i];
+            }
+
+            for i in 1..4 {
+                let d = self.m[i] - self.n[i].to_f64().unwrap();
+                if d >= 1. && self.n[i + 1] - self.n[i] > 1
+                    || d <= -1. && self.n[i - 1] - self.n[i] < -1
+                {
+                    let d = Float::signum(d);
+                    let q_new = self.parabolic(i, d);
+                    if self.q[i - 1] < q_new && q_new < self.q[i + 1] {
+                        self.q[i] = q_new;
+                    } else {
+                        self.q[i] = self.linear(i, d);
+                    }
+                    let delta = d.round() as i64;
+                    debug_assert_eq!(delta.abs(), 1);
+                    self.n[i] += delta;
+                }
+            }
+        }
+
+        pub fn with_samples(mut self, samples: impl IntoIterator<Item = f64>) -> Self {
+            for sample in samples {
+                self.add(sample);
+            }
+
+            self
+        }
+    }
+
+    #[derive(Clone, Copy)]
+    struct FloatOrd(f64);
```

Comment:
Could you give some insights on the ordering that this provides?

Reply:
I added a comment in `fedce6287405fe04826344c3b5537375b317760a`, the ordering is:

`NaN | -Infinity | x < 0 | -0 | +0 | x > 0 | +Infinity | NaN`

It's the same strategy used by [float-ord](https://crates.io/crates/float-ord), which is in turn a dependency of the [average](https://crates.io/crates/average) crate where this quantile computation comes from. These are only used for the dynamic thresholding, so I thought it'd be better to hide the logic within this module versus introducing additional dependencies crate-wide.